zerver/tests/test_message_send.py

```diff
@@ -2768,12 +2768,10 @@ def test_no_topic_message(self) -> None:
         addressee = Addressee.for_stream(stream, topic_name)
 
         do_set_realm_property(realm, "mandatory_topics", True, acting_user=None)
-        realm.refresh_from_db()
 
         with self.assertRaisesRegex(JsonableError, "Topics are required in this organization"):
             check_message(sender, client, addressee, message_content, realm)
```

Comment:
I think here, again, I do want us to do a refresh; calling `check_message(sender, client, addressee, message_content, realm)` means that the test may now be relying on the fact that `do_set_realm_property` mutates the `realm` variable that was passed in.

The concern would not exist if we were instead doing an API call.

Reply:
The whole point is that `do_set_realm_property` does indeed mutate the `realm` variable, so there's no need to refresh it.

Reply:
We're relying on standard Django here.  When you're just testing business logic, you can trust that Django actually persisted stuff to the database without triple-checking that Django did indeed save it by asking it to refresh from the database (which is kind of silly if you don't trust Django in the first place).

You only want to re-fetch DB objects when you're doing stuff like client-server posts where the object clearly gets out of your control.

Reply:
We've had test bugs in the past where a `.save()` was never called in some code path and it resulted in the tests not at all checking the scenario they intended to, or some situation where your local variable was inconsistent with the database. I am sure that at least some of these `refresh_from_db` calls were motivated by a real bug of that form.

Here are a few commits to look at: 685fbffd91e1d93ff76eb423b5ab6f103a786d1b, c78d7288129ddb555b4a4743a57b15e8e33df391; haven't reviewed in full.

We can probably rely on `do_set_realm_property` to not have that class of problem, but I do want to be careful about just blanket switching to cache fetches everywhere.

Reply:
I am not sure about all the cases, but at least for the two commits you mentioned above the object for which we call `refresh_from_db` is not passed to `do_...` function so we need to call `refresh_from_db` for them and I expect this to be the true for most cases where we call `refresh_from_db`.